zbus_macros/src/iface.rs

```diff
@@ -1047,35 +1066,47 @@ fn get_args_from_inputs(
             }
         }
 
-        let args_from_msg = match method_type {
-            MethodType::Property(_) => quote! {
-                #server_arg_decl
+        let (hdr_init, msg_init, signal_emitter_arg_decl, args_decl) = match method_type {
+            MethodType::Property(PropertyType::Getter) => {
+                (quote! {}, quote! {}, quote! {}, quote! {})
+            }
+            MethodType::Property(PropertyType::Setter) => (
+                quote! { let hdr = header.as_ref().unwrap(); },
+                quote! {},
+                quote! { #signal_emitter_arg_decl },
```

Comment:
hmm.. while you're at it, you might as well make signal_emitter also available in the getter? :thinking:

Reply:
I actually wanted and tried to, but it looks like the emitter is taken from the header, and the header will not always be available for getters. Is that a wrong assumption?

Reply:
> hmm.. while you're at it, you might as well make signal_emitter also available in the getter? 🤔

Here's a quick try:

```
test iface_and_proxy ... FAILED

failures:

---- iface_and_proxy stdout ----
thread 'zbus::Connection executor' panicked at zbus/tests/e2e.rs:95:1:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'iface_and_proxy' panicked at zbus/tests/e2e.rs:875:1:
timeout: the function call took 15000 ms. Max time 15000 ms


failures:
    iface_and_proxy

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 15.00s

error: test failed, to rerun pass `-p zbus --test e2e`
```

Reply:
What I mean is that the way the signal emitter i scurrently extracted is (note `match hdr.path()`):
```rust
1042             } else if signal_context || signal_emitter {
1043                 if signal_emitter_arg_decl.is_some() {
1044                     return Err(Error::new_spanned(
1045                         input,
1046                         "There can only be one `signal_emitter` or `signal_context` argument",
1047                     ));
1048                 }
1049
1050                 let signal_context_arg = &input.pat;
1051
1052                 signal_emitter_arg_decl = Some(quote! {
1053                     let #signal_context_arg = match hdr.path() {
1054                         ::std::option::Option::Some(p) => {
1055                             #zbus::object_server::SignalEmitter::new(connection, p).expect("Infallible conversion failed")
1056                         }
1057                         ::std::option::Option::None => {
1058                             let err = #zbus::fdo::Error::UnknownObject("Path Required".into());
1059                             return connection.reply_dbus_error(&hdr, err).await;
1060                         }
1061                     };
1062                 });
```

If there's another way we can extract the emitter there, then we can certainly pass the emitter along to getters as well. Otherwise, a `None` header means no emitter.

Reply:
> I actually wanted and tried to, but it looks like getting the emitter requires a header,

It doesn't need to be. It just needs access to connection and the relevant object path, both of which should be available from contexts where getter is called from.

> What I mean is that the way the signal emitter is currently extracted is (note `match hdr.path()`):

Right, yeah.

---

Looking more into this, I think perhaps your approach here is better. Someone needing signal emissions from a getter are quite small. If anyone comes up with a good use case, we can add it later.

However, we should update the docs to reflect that the signal emitter arg is not available for property getters.

Reply:
> However, we should update the docs to reflect that the signal emitter arg is not available for property getters.

They already do, [zbus_macros/src/lib.rs](https://github.com/dbus2/zbus/pull/1179/commits/1d80d07a1131c25944be2da1fecfef3a62626e45#diff-5110770af24f0b6088a2af69201805b4e31b1f4061ea76c56acd1a9fffc16cca) says (change made in the previous PR): "This argument is not available for property getters."